### PR TITLE
Refactor ApplicabilityActionProvider

### PR DESCRIPTION
**Pull request by Or-Geva**
- [ ] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] I used `npm run format` for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: src/main/diagnostics/abstractFileActionProvider.ts [hunk 1]
```diff
@@ -65,6 +67,17 @@ export abstract class AbstractFileActionProvider implements ExtensionComponent,
...
+        let diagnostic: vscode.Diagnostic = new vscode.Diagnostic(
+            new vscode.Range(issue.regionWithIssue.start, issue.regionWithIssue.end),
+            issue.issueId,
+            vscode.DiagnosticSeverity.Warning
+        );
+        diagnostic.source = this.getSource();
+        diagnostic.code = 'Severity: ' + SeverityUtils.getString(issue.severity);
+        return diagnostic;
+    }
+
```
Comment (attiasas) [inline src/main/diagnostics/abstractFileActionProvider.ts:80]: this is redundant, you can use the createDiagnostics method.
I don't think we should create in the abstract class same method that is more specific for issues with region.
also the creation of the range is also redundant, issue.regionWithIssue is already a range.
I think you should remove this and use the general interface.
we can split the original method to createDiagnostic that only handles one location and return one diagnostic and than use it in the original createDiagnostics if you want to use the single one it in the map method that this method is used for.

File: src/main/treeDataProviders/issuesTree/issuesRootTreeNode.ts [hunk 2]
```diff
@@ -71,6 +71,10 @@ export class IssuesRootTreeNode extends vscode.TreeItem {
         return child;
     }
 
+    public getFileTreeNode(file: string): FileTreeNode | undefined {
+        return this._children.find(child => child.fullPath === file);
+    }
+
```
Comment (attiasas) [inline src/main/treeDataProviders/issuesTree/issuesRootTreeNode.ts:77]: add tests  for this method in issuesTreeNode.test.ts

File: src/main/diagnostics/applicabilityActionProvider.ts [hunk 3]
```diff
@@ -71,45 +71,38 @@ export class ApplicabilityActionProvider extends AbstractFileActionProvider impl
...
+
+    private async addGutterToIssues(document: vscode.TextDocument, issues: CodeIssueTreeNode[]) {
+        let gutters: Map<vscode.Range, Severity> = this.issuesToGutters(issues);
+        const textEditor: vscode.TextEditor = await vscode.window.showTextDocument(document, vscode.ViewColumn.One);
+        for (const [region, severity] of gutters) {
+            this.addGutter(textEditor, SeverityUtils.getIcon(severity), region);
         }
     }
+
+    private issuesToGutters(issues: CodeIssueTreeNode[]): Map<vscode.Range, Severity> {
```
Comment (attiasas) [inline src/main/diagnostics/applicabilityActionProvider.ts:98]: maybe change the name to 'calculateTopSeverityForRegions' ? something that will better describe what the method does.

**Review by attiasas — CHANGES_REQUESTED** [review]
Nice, look at my notes

File: src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts [hunk 4]
```diff
@@ -568,13 +568,22 @@ export class IssuesTreeDataProvider implements vscode.TreeDataProvider<IssuesRoo
...
+            if (this.isWorkspaceContainsFile(workspace.uri.fsPath, filePath)) {
+                return issuesRoot?.getFileTreeNode(filePath);
             }
         }
         return undefined;
     }
 
+    public getCodeIssueTree(filePath: string): CodeFileTreeNode | undefined {
+        const tree: FileTreeNode | undefined = this.getFileIssuesTree(filePath);
+        return tree instanceof CodeIssueTreeNode ? tree : undefined;
```
Comment (attiasas) [inline src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts:580]: tree instance of CodeFileTreeNode not CodeIssueTreeNode

**Review by attiasas — CHANGES_REQUESTED** [review]

File: src/main/diagnostics/applicabilityActionProvider.ts (hunk 3 above)
Comment (Or-Geva) [reply src/main/diagnostics/applicabilityActionProvider.ts:98]: Renamed to 'filterByTopSeverity'

File: src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts (hunk 4 above)
Comment (Or-Geva) [reply src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts:580]: Good catch!

**Pull request merged**